packages/xstate-react/src/useMachine.ts

```diff
@@ -0,0 +1,133 @@
+   * start at this state instead of its `initialState`.
+   */
+  state?: StateConfig<TContext, TEvent>;
+}
+
+export function useMachine<
+  TContext,
+  TEvent extends EventObject,
+  TTypestate extends Typestate<TContext> = any
+>(
+  machine: StateMachine<TContext, any, TEvent, TTypestate>,
+  options: Partial<InterpreterOptions> &
+    Partial<UseMachineOptions<TContext, TEvent>> &
+    Partial<MachineOptions<TContext, TEvent>> = {}
+): [
+  State<TContext, TEvent, any, TTypestate>,
+  Interpreter<TContext, any, TEvent, TTypestate>['send'],
+  Interpreter<TContext, any, TEvent, TTypestate>
+] {
+  if (process.env.NODE_ENV !== 'production') {
+    const [initialMachine] = useState(machine);
+    if (machine !== initialMachine) {
+      console.warn(
+        'Machine given to `useMachine` has changed between renders. This is not supported and might lead to unexpected results.\n' +
+          'Please make sure that you pass the same Machine as argument each time.'
+      );
+    }
+  }
+
+  const {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays,
+    state: rehydratedState,
+    ...interpreterOptions
+  } = options;
+
+  // Capture all actions (side-effects) to be executed.
+  // These will be flushed when they are executed, and avoids the issue of batched events
+  // sent to the interpreter, which might ignore actions.
+  const actionStatesRef = useRef<Array<State<TContext, TEvent>>>([]);
+
+  // Keep a single reference to the invoked machine (the service)
+  const service = useConstant(() => {
+    const machineConfig = {
+      context,
+      guards,
+      actions,
+      activities,
+      services,
+      delays
+    };
+    const createdMachine = machine.withConfig(machineConfig, {
+      ...machine.context,
+      ...context
+    } as TContext);
+
+    // Ensure that actions are not executed (until useEffect() below)
+    interpreterOptions.execute = false;
+
+    return interpret(createdMachine, interpreterOptions).start(
+      rehydratedState ? State.create(rehydratedState) : undefined
+    );
+  });
+
+  // Make sure actions and services are kept updated when they change.
+  // This mutation assignment is safe because the service instance is only used
+  // in one place -- this hook's caller.
+  useEffect(() => {
+    Object.assign(service.machine.options.actions, actions);
+  }, [actions]);
+  useEffect(() => {
+    Object.assign(service.machine.options.services, services);
+  }, [services]);
+
+  // Initialize the state with the initial state.
+  const [state, setState] = useState(service.state);
+
+  useEffect(() => {
+    // Whenever a new state is emitted from the service,
+    // update the state with that state, but only if
+    // that state has changed.
+    service.onTransition((currentState) => {
+      if (currentState.changed) {
+        // capture side-effects to be executed
+        actionStatesRef.current.push(currentState);
```

Comment:
im pretty sure that you miss initial actions (and any actions which happen before useEffect's callback) by starting to queue them from this point in time

Reply:
We can test this, but the initial state/actions are sent to this `onTransition` callback.

Reply:
How could they if you setup onTransition callback in an effect and start a machine during initial render? The initial transitions are getting resolved right after the machine gets started which is before you start listening for transitions

Reply:
The initial state includes actions.

Reply:
The initial state is not pushed onto this `actionStatesRef` though 😉

Reply:
Fixed